Full Vulkan moltenvk support

alexandre-janniaux opened the pull request:
This PR is adding everything needed to compile and run the VulkanRenderer on MacOSX over Metal.

![image](https://user-images.githubusercontent.com/413037/166516063-4981ff69-6550-4c50-a6e6-9758ff907c39.png)

The platform-specific code is written in Objective-C, so it's wrapped into a separate file behind a simple function. If necessary, the wrapper might be extended later to support advanced resizing, color management and events. I haven't tested mouse support also, which might be broken by the current layer properties.

As a side note, it fixes a dependency from `cxxopts` in the Shader module, since it includes it but never actually depended on it before, breaking the compilation on macosx without a global cxxopts.h.

alexandre-janniaux committed:
VulkanRenderer: implement metal view

The metal view backing the MoltenVK compatibility wrapper is written in
Obj-C.

It would have been possible to use the Objective-C runtime in C++ but
the code is less performant (the symbol binding is done at first use
instead of the beginning of the program) and actually harder to get
right.

Note that MoltenVK is not linked to the loader, so the libMoltenVK.dylib
object must be available for loading.

alexandre-janniaux commented:
Note: `#include <objc/runtime.h>` is there for the `id` type from the Objective-C runtime, representing an NSObject that can receive messages.

SirLynix commented on include/Nazara/VulkanRenderer/Wrapper/Surface.inl:
```diff
@@ -6,6 +6,7 @@
 #include <Nazara/Core/Error.hpp>
 #include <Nazara/VulkanRenderer/Utils.hpp>
 #include <Nazara/VulkanRenderer/Debug.hpp>
+#include <vulkan/vulkan_metal.h>
```
Useless as it's already part of Surface.hpp

alexandre-janniaux replied:
Removed, indeed.

SirLynix commented on src/Nazara/VulkanRenderer/VulkanSurface.cpp:
```diff
@@ -7,8 +7,17 @@
 namespace Nz
 {
+	#ifdef VK_USE_PLATFORM_METAL_EXT
+	id CreateAndAttachMetalLayer(void *window);
```
Please use `void* window` instead of `void *window`

alexandre-janniaux replied:
I should have fixed the cases with `type *` and `type &` now.

SirLynix commented on src/Nazara/VulkanRenderer/VulkanSurfaceMetal.mm:
```diff
@@ -0,0 +1,16 @@
+
+namespace Nz
+{
+    id CreateAndAttachMetalLayer(void *window) {
```
same here

SirLynix commented on src/Nazara/VulkanRenderer/VulkanSurfaceMetal.mm:
```diff
@@ -0,0 +1,16 @@
+namespace Nz
+{
+    id CreateAndAttachMetalLayer(void *window) {
+        NSWindow *obj = (__bridge NSWindow*)window;
```
Please use tabs instead of spaces for indentation

SirLynix commented on xmake.lua:
```diff
@@ -81,7 +81,7 @@ local modules = {
 	Shader = {
 		Deps = {"NazaraCore"},
-		Packages = {"efsw", "fmt", "frozen"},
+		Packages = {"cxxopts", "efsw", "fmt", "frozen"},
```
cxxopts is not used by NazaraShader, it's used by nzslc.

alexandre-janniaux replied:
Removed!

SirLynix commented on xmake.lua:
```diff
@@ -103,6 +103,9 @@ local modules = {
 				add_defines("VK_USE_PLATFORM_WAYLAND_KHR")
 			elseif is_plat("macosx") then
 				add_defines("VK_USE_PLATFORM_METAL_EXT")
+				add_links("objc")
```
Prefer `add_syslinks` for system-specific links

alexandre-janniaux replied:
It was needed when I was using the objc runtime directly but it doesn't seem necessary now. I guess xmake can pull it automatically since there is a objc file linked. I'll remove.

SirLynix reviewed (changes requested):
Could you try to enable full engine compilation on the macOS CI? 

In .github/workflows/macos-build.yml, replace:
```
    # Build the engine
    - name: Build Nazara
      run: |
        xmake build -v NazaraPhysics3D
        xmake build -v NazaraWidgets
        xmake build -v NazaraShaderNodes
        xmake build -v PluginAssimp
        xmake build -v NazaraUnitTests
```
by
```
    # Build the engine
    - name: Build Nazara
      run: xmake
```
(you can use `xmake -v` to check the compilation flags if required)

alexandre-janniaux committed:
github: macos-build: build the whole engine

Enable the build of the whole engine now that VulkanRenderer can compile
on macOS.

SirLynix commented:
Thanks a lot!
